modules/libcom/src/misc/cantProceed.c

```diff
@@ -54,6 +54,16 @@ LIBCOM_API void * mallocMustSucceed(size_t size, const char *msg)
 LIBCOM_API void cantProceed(const char *msg, ...)
 {
     va_list pvar;
+#if defined(__GNUC__)
+    static __thread char dead = 0;
+    if(dead) {
+        /* Probably here after a failure in OSD early initialization.
+         * So we don't know which, if any, of the OSI primitives are usable.
+         */
+        abort();
```

Comment:
I assume you're proposing this PR after seeing actual loops. I am unable to trigger a problem by calling `cantProceed()` from the first line of `main()` so it has to be related to some earlier initialization in a static constructor.

My guess at the culprit is `epicsStackTrace()` which is a relatively recent addition to this code and arrived long after we fixed a bunch of OSD Thread Init problems early on, but I don't know if we ever wrote down the rules we had developed to prevent issues. If you can replicate the problem it would be good to check whether commenting out that call resolves it. There is a call path to `epicsMutexMustCreate()` in `epicsStackTrace()` which might explain the loop, although it's from inside a `once()` routine.

Reply:
I hit recursion while messing around with `PTHREAD_PRIO_INHERIT`.  Try calling `errlogPrintf()` from the early initialization code in osdThread or similar.

Reply:
> Try calling `errlogPrintf()` from the early initialization code in osdThread

I can see some problems, we have 4 calls to `checkStatusQuit()` instead of `checkStatusOnceQuit()` in some of the routines in posix/osdThread.c that get called before `epicsThreadInit()` has completed. This comment in the code explains:

```C
/* The following are for use by once, which is only invoked from epicsThreadInit*/
/* Until epicsThreadInit completes errlogInit will not work                     */
/* It must also be used by init_threadInfo otherwise errlogInit could get  */
/* called recursively                                                      */
```
Note that `errlogPrintf()` calls `errlogInit()`.

The initialization strategy in posix/osdThread.c was that API routines must first call `epicsThreadInit()` or otherwise ensure that they are safe to run earlier than that. Some do instead use `assert(epicsThreadOnceCalled)` if they would only be called after some other routine that calls `epicsThreadInit()`. Some newer APIs aren't doing either.